### PR TITLE
Include FontName in Font.GetHashCode

### LINKED ISSUE #44343: System.Drawing.Font.GetHashCode() return is not unique because it does not include the fontName
_This issue has been moved from [a ticket on Developer Community](https://developercommunity2.visualstudio.com/t/SystemDrawingFontGetHashCode-return/998779)._

---
<p>System.Drawing.Font.GetHashCode hashes the fontStyle, fontUnit, and fontSize, but does not hash the fontName.     Thus, New Font(&quot;Arial&quot;, 16, FontStyle.Bold, GraphicsUnit.Point). GetHashCode() returns the same value as New Font(&quot;Lucida Sans Unicode&quot;, 16, FontStyle.Bold, GraphicsUnit.Point). GetHashCode()</p>

---
#### Feedback Bot on 4/20/2020, 08:19 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Feedback Bot on 5/28/2020, 08:00 PM: 

I have detected that for the last 35 days, this issue didn't have much product team activity and a very small amount of new votes or comments. Based on this, its severity, and affected area, it’s my experience that this issue is very unlikely to be fixed.
#### Dada Huo [MSFT] on 11/3/2020, 00:29 AM: 

<p>Thank you for taking the time to log this issue! Did you still reproduce this issue in the latest stable version of Visual Studio installed from <a target='_blank' href="https://visualstudio.microsoft.com/downloads">https://visualstudio.microsoft.com/downloads</a> ?<br>
If so, in order for us to investigate this further, could you please provide repro steps, a more detailed description of the faulty behavior and the actions you were taking when the issue appeared to help us better understand this scenario? It would be very helpful if you could provide some screenshots.<br>
We look forward to hearing from you!</p>

#### Peter Knibbe on 11/5/2020, 07:03 AM: 

<p>Yes. I took the time to build a small project that demonstrates the behavior. Here is the code and I have attached a screenshot.</p>
<p>Public Class Form1<br>
Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click</p>
<pre><code>    Dim fmss = New Font(&quot;Microsoft Sans Serif&quot;, 12)
    Dim bsmt = New Font(&quot;Brush Script MT&quot;, 12)

    Dim match As Boolean = (fmss.GetHashCode() = bsmt.GetHashCode())
    Label3.Text = $&quot;The two fonts {If(match, &quot;&quot;, &quot;do not &quot;)} have the same hash code.&quot;
End Sub
</code></pre>
<p>End Class</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by iSatishYadav**
Fixes: #44343
Added tests for GetHashCode as requested in PR #46816

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=47674) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=47674) before we can accept your contribution.<br/><br/>:x: iSatishYadav [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=47674)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=47674) it.</sub>

File: src/libraries/System.Drawing.Common/tests/FontTests.cs [hunk 1]
```diff
@@ -978,5 +978,18 @@ private static void VerifyFont(Font font, string expectedName, float expectedEmS
             Assert.False(font.IsSystemFont);
             Assert.Empty(font.SystemFontName);
         }
+
+        [ConditionalFact(Helpers.IsDrawingSupported)]
+        public void GetHashCode_DifferentNameSameSizeStyleUnit_HashCodeIsNotSame()
```
Comment (safern) [inline src/libraries/System.Drawing.Common/tests/FontTests.cs:983]: It seems like you need to add
```cs
[SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "GetHashCode doesn't include font name in .NET Framework")]
```

File: src/libraries/System.Drawing.Common/tests/FontTests.cs [hunk 2]
```diff
@@ -978,5 +978,18 @@ private static void VerifyFont(Font font, string expectedName, float expectedEmS
...
+        {
+            using (FontFamily family1 = FontFamily.GenericMonospace)
```
Comment (safern) [inline src/libraries/System.Drawing.Common/tests/FontTests.cs:985]: It seems like RedHat and Fedora are returning the same font family for `GenericMonospace` and `GenericSericf` and that's why the test is failing on those configurations, can we try and use another font family?

File: src/libraries/System.Drawing.Common/tests/FontTests.cs (hunk 1 above)
Comment (iSatishYadav) [reply src/libraries/System.Drawing.Common/tests/FontTests.cs:983]: Oh, that's super helpful @safern. I'll add that.

**Comment by ericstj** [comment]
@iSatishYadav it looks like this is still failing validation, were you planning on coming back to this PR?

**Comment by iSatishYadav** [comment]
> @iSatishYadav it looks like this is still failing validation, were you planning on coming back to this PR?

@ericstj Sorry, I’ll try to work in this this weekend. Could you please suggest any font family which I can use to make this test pass?

**Comment by ericstj** [comment]
You should be able to reproduce the test failure locally to debug.  Looks to me like it fails on Linux, so either use a linux machine or WSL to reproduce.  The build has been deleted due to age so we're missing test results.  I'll queue another.

One thought for fonts to use might be those that were listed in the original issue?  Not sure if those are present on all machines or not.  Another thought might be to just enumerate fonts until you find two that have a different name: `FontFamily.Families`.  I'd wait to see if this is the actual failure though.

**Comment by danmoseley** [comment]
I pushed a potential fix.

**Comment by safern** [comment]
> I pushed a potential fix.

Seems like in Fedora and Redhat is still returning the same font name. I will try and repro locally and see if I find a font that is installed in the docker images that indeed is resolved to a different font name.

**Comment by danmoseley** [comment]
My suggestion is to not spend long on that, and just have the test skip if the font name is the same. It's not platform specific functionality in any way so I don't think it's worth digging into.

**Comment by safern** [comment]
> It's not platform specific functionality in any way so I don't think it's worth digging into.

Makes sense. I can skip if the font name is the same.

**Review by danmoseley — APPROVED** [review]
LGTM!

**Pull request merged**